### PR TITLE
Add SplicePattern AST to parse and type quote pattern splices

### PR DESCRIPTION
**Pull request by nicolasstucki**
Splices in quoted expressions and quoted patterns are fundamentally different. SplicePattern makes this difference explicit and also handles HOAS arguments to the pattern splices.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/Trees.scala [hunk 1]
```diff
@@ -737,6 +737,22 @@ object Trees {
     type ThisTree[+T <: Untyped] = Splice[T]
   }
 
+  /** A tree representing a pattern splice `${ pattern }`, `$ident` or `$ident(args*)` in a quote pattern.
+   *
+   *  Parser will only create `${ pattern }` and `$ident`, hence they will not have args.
+   *  While typing, the `$ident(args*)` the args are identified and desugared into a `SplicePattern`
+   *  containing them.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/ast/Trees.scala:744]: Why not have the parser take care of this since it's based purely on syntactic information?

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/ast/Trees.scala [hunk 2]
```diff
@@ -1566,6 +1587,8 @@ object Trees {
               cpy.Quote(tree)(transform(body)(using quoteContext), transform(tags))
             case tree @ Splice(expr) =>
               cpy.Splice(tree)(transform(expr)(using spliceContext))
+            case tree @ SplicePattern(body, args) =>
+              cpy.SplicePattern(tree)(transform(body)(using spliceContext), transform(args))
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/Trees.scala:1591]: Arguments of the `SplicePattern` are defined and references within the quoted pattern. The definition and references are at the same level.

```scala
case '{ (x: T) => $f(x) } =>
```

In this example `f` is the body, which is at a lower staging level and usable in the RHS of the `=>`. On the other hand `x` is defined in the pattern `(x: T)` and then referenced in the pattern within the `SplicePattern` in `(x)`.

File: compiler/src/dotty/tools/dotc/ast/Trees.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/Trees.scala:744]: In the next PR I will introduce `QuotePattern`. At that point `SplicePattern`s will be kept after parser and until we encode the pattern into `unapply` and `'{..}`s.

File: compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala [hunk 3]
```diff
@@ -74,44 +74,57 @@ trait QuotesAndSplices {
...
+      case tree @ Apply(TypeApply(_, tpt :: Nil), spliced :: Nil) if tree.symbol == defn.QuotedRuntime_exprSplice =>
+        cpy.Splice(tree)(spliced)
+      case tree => tree
+  }
+
+  def typedSplicePattern(tree: untpd.SplicePattern, pt: Type)(using Context): Tree = {
+    record("typedSplicePattern")
+    if (isFullyDefined(pt, ForceDegree.flipBottom)) {
+      def spliceOwner(ctx: Context): Symbol =
+        if (ctx.mode.is(Mode.QuotedPattern)) spliceOwner(ctx.outer) else ctx.owner
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala]: Only in the first iteration. We want to find the first context that is outside of the quote pattern. Note the recursion `spliceOwner(ctx.outer)` and the end `ctx.owner`. I will refactor it to make this clearer.

**Review by smarter — APPROVED** [review]

**Pull request merged**